File: bindings/clojure/loop/io/github/everanium/itb3/clojure/loop/summary.clj
(ns io.github.everanium.itb3.clojure.loop.summary
  "The final summary in both renderings, and the two measurements it
  folds in that are not per-worker counters: the process resident set
  and the shared library's pool counters."
  (:require [clojure.string :as str]
            [io.github.everanium.itb3.clojure.loop.payload :as payload]
            [io.github.everanium.itb3.clojure.loop.size :as size]
            [io.github.everanium.itb3.clojure.loop.state :as state]
            [io.github.everanium.itb3.clojure.loop.worker :as worker]
            [io.github.everanium.itb3.clojure.runtime :as runtime])
  (:import [java.nio.file Files Path]
           [java.util Locale]))

(defn- status-kb
  "Parses one \"Vm...:   1234 kB\" line of /proc/self/status into
  bytes; zero on any parse failure."
  ^long [^String line]
  (let [colon (.indexOf line (int \:))]
    (if (neg? colon)
      0
      (let [first-field (first (str/split (str/trim (subs line (inc colon))) #"\s+"))]
        (try
          (* 1024 (Long/parseLong first-field))
          (catch Exception _ 0))))))

(defn read-rss
  "The process's current resident set and its high-water mark in
  bytes, from /proc/self/status (VmRSS and VmHWM, reported in kB).
  Both are zero on a platform without that file; the figures are
  informational and never enter the verdict."
  []
  (try
    ;; Clojure-specific. The file is read through the NIO line reader
    ;; rather than slurped: a procfs entry reports no length, and the
    ;; stream decoder behind the whole-file reader asks the descriptor
    ;; how many bytes are available before it decodes any.
    (reduce (fn [[current peak] ^String line]
              (cond
                (.startsWith line "VmRSS:") [(status-kb line) peak]
                (.startsWith line "VmHWM:") [current (status-kb line)]
                :else [current peak]))
            [0 0]
            (Files/readAllLines (Path/of "/proc/self/status" (into-array String []))))
    (catch Exception _ [0 0])))

(defn pool-snapshot
  "Pool counters. The shared library keeps process-wide monotonic
  totals at every pool checkout of its cipher core: per hash-array
  tier the starter width, checkouts, constructor misses, regrow
  replacements and bytes allocated; for the scratch byte pool and the
  parallax chunk pool the checkouts, constructor misses, regrows and
  regrow bytes. Two snapshots bracketing the main loop are differenced
  into per-run hit / miss figures that tell whether a pool keeps its
  items warm between calls or evicts them across GC cycles. The slot
  layout is read from the library: slot 0 carries the tier count T,
  tier i occupies the five slots at 1 + 5*i, and the two byte pools
  occupy the eight slots at 1 + 5*T; the vector is sized by the
  binding from the library's own length query, never from a constant.
  Empty when the library is unavailable."
  ^longs []
  (try
    (runtime/pool-stats)
    (catch Exception _ (long-array 0))))

(defn- pool-diff
  "Differences the shutdown snapshot against the warmup baseline into
  the per-tier and per-byte-pool figures the summary reports."
  [^longs steady ^longs warmup]
  (let [empty-delta {:tiers [] :buf {:get 0 :fresh 0 :regrow 0 :regrow-bytes 0}
                     :chunk {:get 0 :fresh 0 :regrow 0 :regrow-bytes 0}}]
    (if (or (< (alength steady) 9) (not= (alength warmup) (alength steady)))
      empty-delta
      (let [tiers (aget steady 0)]
        (if (or (neg? tiers) (> (+ 1 (* 5 tiers) 8) (alength steady)))
          empty-delta
          (let [t (int (+ 1 (* 5 tiers)))]
            {:tiers (into []
                          (comp (map (fn [^long i]
                                       (let [b (int (+ 1 (* 5 i)))]
                                         (when-not (zero? (aget steady b))
                                           {:index i
                                            :starter (aget steady b)
                                            :get (- (aget steady (+ b 1)) (aget warmup (+ b 1)))
                                            :fresh (- (aget steady (+ b 2)) (aget warmup (+ b 2)))
                                            :regrow (- (aget steady (+ b 3)) (aget warmup (+ b 3)))
                                            :new-bytes (- (aget steady (+ b 4))
                                                          (aget warmup (+ b 4)))}))))
                                (remove nil?))
                          (range tiers))
             :buf {:get (- (aget steady t) (aget warmup t))
                   :fresh (- (aget steady (+ t 1)) (aget warmup (+ t 1)))
                   :regrow (- (aget steady (+ t 2)) (aget warmup (+ t 2)))
                   :regrow-bytes (- (aget steady (+ t 3)) (aget warmup (+ t 3)))}
             :chunk {:get (- (aget steady (+ t 4)) (aget warmup (+ t 4)))
                     :fresh (- (aget steady (+ t 5)) (aget warmup (+ t 5)))
                     :regrow (- (aget steady (+ t 6)) (aget warmup (+ t 6)))
                     :regrow-bytes (- (aget steady (+ t 7)) (aget warmup (+ t 7)))}}))))))

(defn- miss-percent
  "Misses over checkouts as a percentage; zero when nothing was
  checked out."
  ^double [^long miss ^long get-count]
  (if (<= get-count 0)
    0.0
    (/ (* 100.0 miss) get-count)))

(defn- json-string
  "Renders s as a JSON string literal with the escapes JSON requires."
  ^String [^String s]
  (let [sb (StringBuilder. (+ 2 (.length s)))]
    (.append sb \")
    (dotimes [i (.length s)]
      (let [ch (.charAt s i)]
        (cond
          (= ch \") (.append sb "\\\"")
          (= ch \\) (.append sb "\\\\")
          (= ch \newline) (.append sb "\\n")
          (= ch \return) (.append sb "\\r")
          (= ch \tab) (.append sb "\\t")
          (< (int ch) 32) (.append sb (String/format Locale/ROOT "\\u%04x"
                                                     (object-array [(int ch)])))
          :else (.append sb ch))))
    (.append sb \")
    (.toString sb)))

(defn- effective-gogc
  "The effective GC percentage as the runtime reports it: the query
  form of the setter (a set-and-restore round trip inside the library)
  so the field is the same whether the value came from the flag, the
  environment, or the runtime default."
  ^long [^long flag]
  (if (pos? flag)
    flag
    (try
      (runtime/set-gc-percent! -1)
      (catch Exception _ 0))))

(defn emit-summary
  "Output contract. Both renderings are shared with the Go harness and
  every other binding's loop utility field for field: the same lines
  in the same order, the same keys in the same order, floats with a
  fixed number of decimals so the JSON is byte-identical across
  implementations. The Go harness alone adds its runtime-internal
  lines after rss: and its runtime-internal keys after
  parallax_chunk_pool; nothing here reproduces them because nothing
  they read is reachable through the C ABI. Returns the exit code."
  ^long [r ^long elapsed-ns m]
  (let [cfg (:cfg r)
        workers (long (:workers cfg))
        per-worker (mapv #(state/counter-value % :iters) (:counters r))
        total-iters (reduce + 0 per-worker)
        total-enc (reduce + 0 (map #(state/counter-value % :bytes-enc) (:counters r)))
        total-dec (reduce + 0 (map #(state/counter-value % :bytes-dec) (:counters r)))
        nanos-enc (reduce + 0 (map #(state/counter-value % :nanos-enc) (:counters r)))
        nanos-dec (reduce + 0 (map #(state/counter-value % :nanos-dec) (:counters r)))
        errors (into [] (keep #(deref (:error %))) (:counters r))
        ;; Throughput. Per-direction throughput divides the sum of
        ;; every worker's wall time in that direction by the worker
        ;; count — the equivalent single-stream wall time under N-way
        ;; concurrency — so each direction reports the aggregate rate
        ;; it sustained rather than collapsing to combined/2 (every
        ;; iteration moves equal encrypt and decrypt bytes, so a
        ;; total-elapsed denominator would give both directions the
        ;; same figure). The combined rate keeps total elapsed as the
        ;; one-glance overall figure.
        avg-enc (if (pos? (long nanos-enc)) (quot (long nanos-enc) workers) 0)
        avg-dec (if (pos? (long nanos-dec)) (quot (long nanos-dec) workers) 0)
        rss-warmup (long (:rss-warmup m))
        rss-final (long (:rss-final m))
        rss-peak (long (:rss-peak m))
        rss-delta (- rss-final rss-warmup)
        rss-growth (if (pos? rss-warmup) (/ (* 100.0 rss-delta) rss-warmup) 0.0)
        pd (pool-diff (:pool-steady m) (:pool-warmup m))
        pass? (empty? errors)
        rekeys (state/counter-value r :rekeys)
        cycles (state/counter-value r :blob-cycles)
        gomaxprocs (try (runtime/set-gomaxprocs! 0) (catch Exception _ 0))
        pipes @(:pipes r)
        stream-profile (if (:stream pipes) (:stream-profile r) "")
        msg-profile (if (:msg pipes) (:msg-profile r) "")]
    (if (:json-output cfg)
      (let [j (StringBuilder.)]
        (doto j
          (.append "{\"duration_seconds\":") (.append (size/f (/ elapsed-ns 1e9) 3))
          (.append ",\"iterations\":") (.append (long total-iters))
          (.append ",\"per_worker_iterations\":[")
          (.append (str/join "," per-worker)) (.append \])
          (.append ",\"bytes_encrypted\":") (.append (long total-enc))
          (.append ",\"bytes_decrypted\":") (.append (long total-dec))
          (.append ",\"encrypt_mb_per_sec\":")
          (.append (size/f (size/mb-per-sec total-enc avg-enc) 1))
          (.append ",\"decrypt_mb_per_sec\":")
          (.append (size/f (size/mb-per-sec total-dec avg-dec) 1))
          (.append ",\"combined_mb_per_sec\":")
          (.append (size/f (size/mb-per-sec (+ (long total-enc) (long total-dec)) elapsed-ns) 1))
          (.append ",\"rekeys\":") (.append rekeys)
          (.append ",\"blob_cycles\":") (.append cycles)
          (.append ",\"worker_errors\":[")
          (.append (str/join "," (map json-string errors))) (.append \])
          (.append ",\"verdict\":\"") (.append (if pass? "PASS" "FAIL")) (.append \")
          (.append ",\"shape\":\"") (.append ^String (worker/shape-label (:shape cfg)))
          (.append \")
          (.append ",\"stream_profile\":") (.append (json-string stream-profile))
          (.append ",\"message_profile\":") (.append (json-string msg-profile))
          (.append ",\"hash\":") (.append (json-string (:hash cfg)))
          (.append ",\"mac\":") (.append (json-string (:mac cfg)))
          (.append ",\"payload_bytes\":") (.append (long (:payload cfg)))
          (.append ",\"payload_mode\":\"")
          (.append ^String (payload/mode-label (:payload-mode cfg))) (.append \")
          (.append ",\"seed\":") (.append (Long/toUnsignedString (long (:seed cfg))))
          (.append ",\"key_bits\":") (.append (long (:key-bits cfg)))
          (.append ",\"nonce_bits\":") (.append (long (:nonce-bits cfg)))
          (.append ",\"chunk_size_bytes\":") (.append (long (:chunk-size cfg)))
          (.append ",\"barrier_fill\":") (.append (long (:barrier-fill cfg)))
          (.append ",\"parallax\":\"") (.append (state/on-off (:parallax cfg))) (.append \")
          (.append ",\"wrapper\":\"") (.append (state/on-off (:wrapper cfg))) (.append \")
          (.append ",\"goroutines_requested\":") (.append (long (:workers-requested cfg)))
          (.append ",\"goroutines\":") (.append workers)
          (.append ",\"concurrency\":\"") (.append ^String state/concurrency) (.append \")
          (.append ",\"gogc\":\"") (.append (effective-gogc (long (:gogc cfg)))) (.append \")
          (.append ",\"memlimit_bytes\":") (.append (long (:memlimit cfg)))
          (.append ",\"gomaxprocs\":") (.append (long gomaxprocs))
          (.append ",\"microbatch_tiers\":")
          (.append (json-string (state/policy-label "ITB_MICROBATCH_TIERS")))
          (.append ",\"hashpool_starters\":")
          (.append (json-string (state/policy-label "ITB_HASHPOOL_STARTERS")))
          (.append ",\"rss_warmup_bytes\":") (.append rss-warmup)
          (.append ",\"rss_peak_bytes\":") (.append rss-peak)
          (.append ",\"rss_final_bytes\":") (.append rss-final)
          (.append ",\"rss_growth_percent\":") (.append (size/f rss-growth 2))
          (.append ",\"hash_pool_tiers\":[")
          (.append (str/join ","
                             (map (fn [t]
                                    (str "{\"tier\":" (:index t)
                                         ",\"starter\":" (:starter t)
                                         ",\"get\":" (:get t)
                                         ",\"new\":" (:fresh t)
                                         ",\"regrow\":" (:regrow t)
                                         ",\"new_bytes\":" (:new-bytes t)
                                         ",\"miss_percent\":"
                                         (size/f (miss-percent (+ (long (:fresh t))
                                                                  (long (:regrow t)))
                                                               (:get t))
                                                 2)
                                         "}"))
                                  (:tiers pd))))
          (.append \])
          (.append ",\"buf_pool\":{\"get\":") (.append (long (:get (:buf pd))))
          (.append ",\"new\":") (.append (long (:fresh (:buf pd))))
          (.append ",\"regrow\":") (.append (long (:regrow (:buf pd))))
          (.append ",\"regrow_bytes\":") (.append (long (:regrow-bytes (:buf pd))))
          (.append ",\"miss_percent\":")
          (.append (size/f (miss-percent (:regrow (:buf pd)) (:get (:buf pd))) 2))
          (.append \})
          (.append ",\"parallax_chunk_pool\":{\"get\":") (.append (long (:get (:chunk pd))))
          (.append ",\"new\":") (.append (long (:fresh (:chunk pd))))
          (.append ",\"regrow\":") (.append (long (:regrow (:chunk pd))))
          (.append ",\"regrow_bytes\":") (.append (long (:regrow-bytes (:chunk pd))))
          (.append ",\"miss_percent\":")
          (.append (size/f (miss-percent (:regrow (:chunk pd)) (:get (:chunk pd))) 2))
          (.append \})
          (.append \}))
        (println (.toString j))
        (if pass? 0 1))
      (do
        (state/log-line "=== FINAL ===")
        (state/log-line (str "  duration: " (size/human-duration (size/round-to elapsed-ns 1000000))))
        (state/log-line (str "  iterations: " (str/join " + " per-worker)
                             " = " total-iters " total"))
        (state/log-line (str "  throughput: encrypt " (size/human-rate total-enc avg-enc)
                             ", decrypt " (size/human-rate total-dec avg-dec)
                             ", combined " (size/human-rate (+ (long total-enc) (long total-dec))
                                                            elapsed-ns)))
        (state/log-line (str "  bytes: " (size/human-bytes total-enc) " encrypted, "
                             (size/human-bytes total-dec) " decrypted"))
        (state/log-line (str "  data integrity: " total-iters "/" total-iters " PASS"))
        (state/log-line (str "  concurrency: " state/concurrency
                             ", workers " workers
                             " (requested " (:workers-requested cfg) ")"))
        (state/log-line (str "  rss: warmup " (size/human-bytes rss-warmup)
                             ", peak " (size/human-bytes rss-peak)
                             ", final " (size/human-bytes rss-final)
                             " (delta " (size/human-bytes-signed rss-delta)
                             ", " (size/f rss-growth 1) "% growth)"))
        (doseq [t (:tiers pd)]
          (state/log-line (str "  hash pool tier " (:index t)
                               " (starter " (:starter t) "): get " (:get t)
                               ", miss " (+ (long (:fresh t)) (long (:regrow t)))
                               " (new " (:fresh t) " + regrow " (:regrow t) ")"
                               ", miss " (size/f (miss-percent (+ (long (:fresh t))
                                                                  (long (:regrow t)))
                                                               (:get t))
                                                 2)
                               "%, " (size/human-bytes (:new-bytes t)) " allocated")))
        (state/log-line (str "  buf pool: get " (:get (:buf pd))
                             ", regrow " (:regrow (:buf pd))
                             " (of which fresh " (:fresh (:buf pd)) ")"
                             ", miss " (size/f (miss-percent (:regrow (:buf pd))
                                                             (:get (:buf pd)))
                                               2)
                             "%, " (size/human-bytes (:regrow-bytes (:buf pd))) " regrown"))
        (state/log-line (str "  parallax chunk pool: get " (:get (:chunk pd))
                             ", regrow " (:regrow (:chunk pd))
                             " (of which fresh " (:fresh (:chunk pd)) ")"
                             ", miss " (size/f (miss-percent (:regrow (:chunk pd))
                                                             (:get (:chunk pd)))
                                               2)
                             "%, " (size/human-bytes (:regrow-bytes (:chunk pd))) " regrown"))
        (when (pos? rekeys)
          (state/log-line (str "  rekeys: " rekeys)))
        (when (pos? cycles)
          (state/log-line (str "  blob cycles: " cycles)))
        (doseq [e errors]
          (state/log-line (str "  ERROR: " e)))
        (if pass?
          (do (state/log-line "  verdict: PASS") 0)
          (do (state/log-line (str "  verdict: FAIL (errors=" (count errors) ")")) 1))))))

File: bindings/scala/loop/src/main/scala/io/github/everanium/itb3/scala/loop/Summary.scala
// The final summary in both renderings, and the two measurements it
// folds in that are not per-worker counters: the process resident set
// and the shared library's pool counters.

package io.github.everanium.itb3.scala.loop

import java.util.Locale
import scala.io.Source
import scala.util.Using

import io.github.everanium.itb3.scala.Runtime as ItbRuntime

object Summary:

  /** Parses one "Vm...:   1234 kB" line of /proc/self/status into
    * bytes; zero on any parse failure.
    */
  private def statusKb(line: String): Long =
    val colon = line.indexOf(':')
    if colon < 0 then return 0L
    val parts = line.substring(colon + 1).trim.split("\\s+")
    if parts.isEmpty || parts(0).isEmpty then return 0L
    try parts(0).toLong * 1024L
    catch case _: NumberFormatException => 0L

  /** The process's current resident set and its high-water mark in
    * bytes, from /proc/self/status (VmRSS and VmHWM, reported in kB).
    * Both are zero on a platform without that file; the figures are
    * informational and never enter the verdict.
    */
  def readRss(): (Long, Long) =
    val lines =
      try Using.resource(Source.fromFile("/proc/self/status"))(_.getLines().toList)
      catch case _: Exception => return (0L, 0L)
    var current = 0L
    var peak = 0L
    for line <- lines do
      if line.startsWith("VmRSS:") then current = statusKb(line)
      else if line.startsWith("VmHWM:") then peak = statusKb(line)
    (current, peak)

  /** Pool counters. The shared library keeps process-wide monotonic
    * totals at every pool checkout of its cipher core: per hash-array
    * tier the starter width, checkouts, constructor misses, regrow
    * replacements and bytes allocated; for the scratch byte pool and
    * the parallax chunk pool the checkouts, constructor misses, regrows
    * and regrow bytes. Two snapshots bracketing the main loop are
    * differenced into per-run hit / miss figures that tell whether a
    * pool keeps its items warm between calls or evicts them across GC
    * cycles. The slot layout is read from the library: slot 0 carries
    * the tier count T, tier i occupies the five slots at 1 + 5*i, and
    * the two byte pools occupy the eight slots at 1 + 5*T; the vector
    * is sized by the binding from the library's own length query, never
    * from a constant. Empty when the library is unavailable.
    */
  def poolSnapshot(): Array[Long] =
    ItbRuntime.poolStats() match
      case Right(v) => v
      case Left(_)  => Array.empty

  /** One starter tier of the hash-array pool, differenced. */
  private final case class Tier(
      index: Long,
      starter: Long,
      get: Long,
      fresh: Long,
      regrow: Long,
      newBytes: Long
  )

  /** One single-size byte pool, differenced. */
  private final case class BytePool(
      get: Long = 0,
      fresh: Long = 0,
      regrow: Long = 0,
      regrowBytes: Long = 0
  )

  private final case class PoolDelta(
      tiers: List[Tier] = Nil,
      buf: BytePool = BytePool(),
      chunk: BytePool = BytePool()
  )

  private def poolDiff(steady: Array[Long], warmup: Array[Long]): PoolDelta =
    if steady.length < 9 || warmup.length != steady.length then return PoolDelta()
    val tiers = steady(0)
    if tiers < 0 || 1 + 5 * tiers + 8 > steady.length then return PoolDelta()
    val list = scala.collection.mutable.ListBuffer[Tier]()
    var i = 0L
    while i < tiers do
      val b = (1 + 5 * i).toInt
      if steady(b) != 0L then
        list += Tier(
          index = i,
          starter = steady(b),
          get = steady(b + 1) - warmup(b + 1),
          fresh = steady(b + 2) - warmup(b + 2),
          regrow = steady(b + 3) - warmup(b + 3),
          newBytes = steady(b + 4) - warmup(b + 4)
        )
      i += 1
    val t = (1 + 5 * tiers).toInt
    PoolDelta(
      tiers = list.toList,
      buf = BytePool(
        steady(t) - warmup(t),
        steady(t + 1) - warmup(t + 1),
        steady(t + 2) - warmup(t + 2),
        steady(t + 3) - warmup(t + 3)
      ),
      chunk = BytePool(
        steady(t + 4) - warmup(t + 4),
        steady(t + 5) - warmup(t + 5),
        steady(t + 6) - warmup(t + 6),
        steady(t + 7) - warmup(t + 7)
      )
    )

  /** Misses over checkouts as a percentage; zero when nothing was
    * checked out.
    */
  private def missPercent(miss: Long, get: Long): Double =
    if get <= 0 then 0.0 else 100.0 * miss / get

  /** Renders s as a JSON string literal with the escapes JSON
    * requires.
    */
  private def jsonString(s: String): String =
    val sb = StringBuilder()
    sb.append('"')
    for ch <- s do
      ch match
        case '"'                 => sb.append("\\\"")
        case '\\'                => sb.append("\\\\")
        case '\n'                => sb.append("\\n")
        case '\r'                => sb.append("\\r")
        case '\t'                => sb.append("\\t")
        case c if c < ' '        => sb.append(String.format(Locale.ROOT, "\\u%04x", c.toInt))
        case c                   => sb.append(c)
    sb.append('"')
    sb.toString

  /** The effective GC percentage as the runtime reports it: the query
    * form of the setter (a set-and-restore round trip inside the
    * library) so the field is the same whether the value came from the
    * flag, the environment, or the runtime default.
    */
  private def effectiveGogc(flag: Int): Int =
    if flag > 0 then flag else ItbRuntime.setGCPercent(-1)

  /** Output contract. Both renderings are shared with the Go harness
    * and every other binding's loop utility field for field: the same
    * lines in the same order, the same keys in the same order, floats
    * with a fixed number of decimals so the JSON is byte-identical
    * across implementations. The Go harness alone adds its
    * runtime-internal lines after rss: and its runtime-internal keys
    * after parallax_chunk_pool; nothing here reproduces them because
    * nothing they read is reachable through the C ABI. Returns the exit
    * code.
    */
  def emit(r: RunState, elapsedNs: Long): Int =
    val cfg = r.cfg
    val workers = cfg.workers.toLong
    var totalIters = 0L
    var totalEnc = 0L
    var totalDec = 0L
    var nanosEnc = 0L
    var nanosDec = 0L
    val perWorker = scala.collection.mutable.ListBuffer[Long]()
    val errors = scala.collection.mutable.ListBuffer[String]()
    for c <- r.workers do
      val n = c.iters.get()
      perWorker += n
      totalIters += n
      totalEnc += c.bytesEnc.get()
      totalDec += c.bytesDec.get()
      nanosEnc += c.nanosEnc.get()
      nanosDec += c.nanosDec.get()
      c.error.foreach(errors += _)

    // Throughput. Per-direction throughput divides the sum of every
    // worker's wall time in that direction by the worker count — the
    // equivalent single-stream wall time under N-way concurrency — so
    // each direction reports the aggregate rate it sustained rather
    // than collapsing to combined/2 (every iteration moves equal
    // encrypt and decrypt bytes, so a total-elapsed denominator would
    // give both directions the same figure). The combined rate keeps
    // total elapsed as the one-glance overall figure.
    val avgEnc = if nanosEnc > 0 then nanosEnc / workers else 0L
    val avgDec = if nanosDec > 0 then nanosDec / workers else 0L

    val rssDelta = r.rssFinal - r.rssWarmup
    val rssGrowth = if r.rssWarmup > 0 then 100.0 * rssDelta / r.rssWarmup else 0.0

    val pd = poolDiff(r.poolSteady, r.poolWarmup)
    val pass = errors.isEmpty
    val rekeys = r.rekeys.get()
    val cycles = r.blobCycles.get()
    val gomaxprocs = ItbRuntime.setGOMAXPROCS(0)
    val streamProfile = if r.pipes.stream.isDefined then r.streamProfile else ""
    val msgProfile = if r.pipes.msg.isDefined then r.msgProfile else ""

    if cfg.jsonOutput then
      val j = StringBuilder()
      j.append("{\"duration_seconds\":").append(Size.f(elapsedNs / 1e9, 3))
      j.append(",\"iterations\":").append(totalIters)
      j.append(",\"per_worker_iterations\":[").append(perWorker.mkString(",")).append(']')
      j.append(",\"bytes_encrypted\":").append(totalEnc)
      j.append(",\"bytes_decrypted\":").append(totalDec)
      j.append(",\"encrypt_mb_per_sec\":").append(Size.f(Size.mbPerSec(totalEnc, avgEnc), 1))
      j.append(",\"decrypt_mb_per_sec\":").append(Size.f(Size.mbPerSec(totalDec, avgDec), 1))
      j.append(",\"combined_mb_per_sec\":")
        .append(Size.f(Size.mbPerSec(totalEnc + totalDec, elapsedNs), 1))
      j.append(",\"rekeys\":").append(rekeys)
      j.append(",\"blob_cycles\":").append(cycles)
      j.append(",\"worker_errors\":[").append(errors.map(jsonString).mkString(",")).append(']')
      j.append(",\"verdict\":\"").append(if pass then "PASS" else "FAIL").append('"')
      j.append(",\"shape\":\"").append(cfg.shape.label).append('"')
      j.append(",\"stream_profile\":").append(jsonString(streamProfile))
      j.append(",\"message_profile\":").append(jsonString(msgProfile))
      j.append(",\"hash\":").append(jsonString(cfg.hash))
      j.append(",\"mac\":").append(jsonString(cfg.mac))
      j.append(",\"payload_bytes\":").append(cfg.payload)
      j.append(",\"payload_mode\":\"").append(cfg.payloadMode.label).append('"')
      j.append(",\"seed\":").append(java.lang.Long.toUnsignedString(cfg.seed))
      j.append(",\"key_bits\":").append(cfg.keyBits)
      j.append(",\"nonce_bits\":").append(cfg.nonceBits)
      j.append(",\"chunk_size_bytes\":").append(cfg.chunkSize)
      j.append(",\"barrier_fill\":").append(cfg.barrierFill)
      j.append(",\"parallax\":\"").append(Main.onOff(cfg.parallax)).append('"')
      j.append(",\"wrapper\":\"").append(Main.onOff(cfg.wrapper)).append('"')
      j.append(",\"goroutines_requested\":").append(cfg.workersRequested)
      j.append(",\"goroutines\":").append(cfg.workers)
      j.append(",\"concurrency\":\"").append(Main.Concurrency).append('"')
      j.append(",\"gogc\":\"").append(effectiveGogc(cfg.gogc)).append('"')
      j.append(",\"memlimit_bytes\":").append(cfg.memlimit)
      j.append(",\"gomaxprocs\":").append(gomaxprocs)
      j.append(",\"microbatch_tiers\":").append(jsonString(Main.policyLabel("ITB_MICROBATCH_TIERS")))
      j.append(",\"hashpool_starters\":")
        .append(jsonString(Main.policyLabel("ITB_HASHPOOL_STARTERS")))
      j.append(",\"rss_warmup_bytes\":").append(r.rssWarmup)
      j.append(",\"rss_peak_bytes\":").append(r.rssPeak)
      j.append(",\"rss_final_bytes\":").append(r.rssFinal)
      j.append(",\"rss_growth_percent\":").append(Size.f(rssGrowth, 2))
      j.append(",\"hash_pool_tiers\":[")
      j.append(
        pd.tiers
          .map(t =>
            s"{\"tier\":${t.index},\"starter\":${t.starter},\"get\":${t.get}," +
              s"\"new\":${t.fresh},\"regrow\":${t.regrow},\"new_bytes\":${t.newBytes}," +
              s"\"miss_percent\":${Size.f(missPercent(t.fresh + t.regrow, t.get), 2)}}"
          )
          .mkString(",")
      )
      j.append(']')
      j.append(",\"buf_pool\":{\"get\":").append(pd.buf.get)
        .append(",\"new\":").append(pd.buf.fresh)
        .append(",\"regrow\":").append(pd.buf.regrow)
        .append(",\"regrow_bytes\":").append(pd.buf.regrowBytes)
        .append(",\"miss_percent\":").append(Size.f(missPercent(pd.buf.regrow, pd.buf.get), 2))
        .append('}')
      j.append(",\"parallax_chunk_pool\":{\"get\":").append(pd.chunk.get)
        .append(",\"new\":").append(pd.chunk.fresh)
        .append(",\"regrow\":").append(pd.chunk.regrow)
        .append(",\"regrow_bytes\":").append(pd.chunk.regrowBytes)
        .append(",\"miss_percent\":").append(Size.f(missPercent(pd.chunk.regrow, pd.chunk.get), 2))
        .append('}')
      j.append('}')
      println(j.toString)
      return if pass then 0 else 1

    Main.logLine("=== FINAL ===")
    Main.logLine("  duration: " + Size.humanDuration(Size.roundTo(elapsedNs, 1_000_000L)))
    Main.logLine("  iterations: " + perWorker.mkString(" + ") + " = " + totalIters + " total")
    Main.logLine(
      "  throughput: encrypt " + Size.humanRate(totalEnc, avgEnc) +
        ", decrypt " + Size.humanRate(totalDec, avgDec) +
        ", combined " + Size.humanRate(totalEnc + totalDec, elapsedNs)
    )
    Main.logLine(
      "  bytes: " + Size.humanBytes(totalEnc) + " encrypted, " +
        Size.humanBytes(totalDec) + " decrypted"
    )
    Main.logLine("  data integrity: " + totalIters + "/" + totalIters + " PASS")
    Main.logLine(
      "  concurrency: " + Main.Concurrency + ", workers " + cfg.workers +
        " (requested " + cfg.workersRequested + ")"
    )
    Main.logLine(
      "  rss: warmup " + Size.humanBytes(r.rssWarmup) +
        ", peak " + Size.humanBytes(r.rssPeak) +
        ", final " + Size.humanBytes(r.rssFinal) +
        " (delta " + Size.humanBytesSigned(rssDelta) +
        ", " + Size.f(rssGrowth, 1) + "% growth)"
    )
    for t <- pd.tiers do
      Main.logLine(
        "  hash pool tier " + t.index + " (starter " + t.starter + "): get " + t.get +
          ", miss " + (t.fresh + t.regrow) +
          " (new " + t.fresh + " + regrow " + t.regrow + ")" +
          ", miss " + Size.f(missPercent(t.fresh + t.regrow, t.get), 2) + "%" +
          ", " + Size.humanBytes(t.newBytes) + " allocated"
      )
    Main.logLine(
      "  buf pool: get " + pd.buf.get + ", regrow " + pd.buf.regrow +
        " (of which fresh " + pd.buf.fresh + ")" +
        ", miss " + Size.f(missPercent(pd.buf.regrow, pd.buf.get), 2) + "%" +
        ", " + Size.humanBytes(pd.buf.regrowBytes) + " regrown"
    )
    Main.logLine(
      "  parallax chunk pool: get " + pd.chunk.get + ", regrow " + pd.chunk.regrow +
        " (of which fresh " + pd.chunk.fresh + ")" +
        ", miss " + Size.f(missPercent(pd.chunk.regrow, pd.chunk.get), 2) + "%" +
        ", " + Size.humanBytes(pd.chunk.regrowBytes) + " regrown"
    )
    if rekeys > 0 then Main.logLine("  rekeys: " + rekeys)
    if cycles > 0 then Main.logLine("  blob cycles: " + cycles)
    for e <- errors do Main.logLine("  ERROR: " + e)
    if pass then
      Main.logLine("  verdict: PASS")
      0
    else
      Main.logLine("  verdict: FAIL (errors=" + errors.size + ")")
      1
